src/s3/util/fileViewerManager.ts

```diff
@@ -21,25 +21,73 @@ import { localize } from '../../shared/utilities/vsCodeUtils'
 const SIZE_LIMIT = 4 * Math.pow(10, 6)
 
 export class S3FileViewerManager {
-    private cacheArns: Set<string>
-    private activeTabs: Map<string, S3Tab>
-    private window: typeof vscode.window
     private outputChannel: OutputChannel
-    private commands: Commands
-    private tempLocation: string | undefined
+    private promptOnEdit = true
+    //onDidChange to trigger refresh of contents on the document provider
+    private _onDidChange = new vscode.EventEmitter<vscode.Uri>()
+    public get onDidChange(): vscode.Event<vscode.Uri> {
+        return this._onDidChange.event
+    }
 
     public constructor(
-        cacheArn: Set<string> = new Set<string>(),
-        window: typeof vscode.window = vscode.window,
-        commands = Commands.vscode(),
-        tempLocation?: string
+        private cacheArns: Set<string> = new Set<string>(),
+        private window: typeof vscode.window = vscode.window,
+        private commands = Commands.vscode(),
+        private _tempLocation?: string,
+        private activeTabs: Map<string, S3Tab> = new Map<string, S3Tab>()
     ) {
-        this.cacheArns = cacheArn
-        this.activeTabs = new Map<string, S3Tab>()
-        this.window = window
-        this.commands = commands
-        this.tempLocation = tempLocation
         this.outputChannel = ext.outputChannel
+
+        vscode.workspace.onDidSaveTextDocument(async savedTextDoc => {
+            if (this.activeTabs.has(savedTextDoc.uri.fsPath)) {
+                const activeTab = this.activeTabs.get(savedTextDoc.uri.fsPath)!
+
+                if (!(await this.checkForValidity(activeTab.s3FileNode, activeTab.fileUri))) {
+                    window.showErrorMessage(
+                        localize(
+                            'AWS.s3.fileViewer.error.invalidUpload',
+                            'File is invalid to upload, file has changed in S3 since last cache download. Please try and reopen the file. Be aware current changes may be lost.'
+                        )
+                    )
+                    return
+                }
+
+                if (!(await activeTab.uploadChangesToS3())) {
+                    this.window.showErrorMessage(
+                        'Error uploading file to S3. Changes will not be saved. Please try and resave this edit mode file'
```

Comment:
Unless you can force VS Code to not write the file to the system, don't say that the change won't be saved.

Reply:
'Changes will not be saved to S3' maybe?